File: BST.js
class Node {
  constructor(val) {
  this.val = val
  this.left = null
  this.right = null
  }
}
class BinarySearchTree {
  constructor() {
  this.root = null
  }
  push(val) {
      if(!this.root) {
        console.log("Inside push");
        this.root = new Node(val)
        return
      }
        var current = this.root
        var newNode = new Node(val)
        while(current){
          console.log("while");
            if(current.val>val) {
              if(current.left)
               current = current.left
              else {
                current.left = newNode
                break
              }
            }
            else {
              if(current.right)
               current = current.right
              else {
                current.right = newNode
                break
              }
            }
        }
    }
      printInOrder(root) {
        console.log("Print "+root.data);
        if(root!=null) {
          bst.printInOrder(root.left);
          console.log(root.data);
          bst.printInOrder(root.right);
        }
      }
  }
  var bst = new BinarySearchTree();
  bst.push(8);
  bst.push(2);
  bst.push(4);
  bst.push(15);
  bst.push(10);
  bst.push(9);
  bst.push(14);
  bst.push(24);
  bst.push(18);
  bst.push(60);
  bst.printInOrder(bst.root);

File: BST.java
/* Hidden stub code will pass a root argument to the function below. Complete the function to solve the challenge. Hint: you may want to write one or more helper functions.

The Node class is defined as follows:
    class Node {
        int data;
        Node left;
        Node right;
     }
*/
    int maxInLeft(Node leftData) {
        int max = -1;
        while(leftData!=null) {
            if(leftData.data>max)
                max = leftData.data;
            leftData = leftData.left;
        }
        return max;
    }
    int minInRight(Node rightData) {
      int min = 100000;
       while(rightData!=null) {
           if(rightData.data<min)
               min = rightData.data;
           rightData = rightData.right;
       }
        return min;
    }
    boolean checkBST(Node root) {
      if(root==null) {
          return true;
      }
      if(root.left!=null && maxInLeft(root.left)>root.data)
        return false;
      if(root.left!=null && maxInLeft(root.left)==root.data)
        return false;
      if(root.right!=null && minInRight(root.right)<root.data)
          return false;
      if(root.right!=null && minInRight(root.right)==root.data)
          return false;
      if(checkBST(root.right)==false || checkBST(root.left)==false)
          return false;
      return true;
    }
